### ISSUE TITLE
regression: visibility qualifiers are not permitted

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#121607]
* https://crater-reports.s3.amazonaws.com/beta-1.77-3/beta-2024-02-18/reg/amadeus-core-0.4.3/log.txt
  * https://crater-reports.s3.amazonaws.com/beta-1.77-3/beta-2024-02-18/reg/amadeus-aws-0.4.3/log.txt
  * https://crater-reports.s3.amazonaws.com/beta-1.77-3/beta-2024-02-18/reg/amadeus-commoncrawl-0.4.3/log.txt
  * https://crater-reports.s3.amazonaws.com/beta-1.77-3/beta-2024-02-18/reg/amadeus-parquet-0.4.3/log.txt
  * https://crater-reports.s3.amazonaws.com/beta-1.77-3/beta-2024-02-18/reg/amadeus-postgres-0.4.3/log.txt
  * https://crater-reports.s3.amazonaws.com/beta-1.77-3/beta-2024-02-18/reg/amadeus-serde-0.4.3/log.txt
  * https://crater-reports.s3.amazonaws.com/beta-1.77-3/beta-2024-02-18/reg/amadeus-types-0.4.3/log.txt
 

```
[INFO] [stdout] error[E0449]: visibility qualifiers are not permitted here
[INFO] [stdout]    --> src/par_stream.rs:647:16
[INFO] [stdout]     |
[INFO] [stdout] 647 |                   pool.spawn(FnOnce!(move |pool: &P::ThreadPool| {
[INFO] [stdout]     |  ____________________________^
[INFO] [stdout] 648 | |                     let mut process_tasks = tasks.into_iter();
[INFO] [stdout] 649 | |
[INFO] [stdout] 650 | |                     let mut tasks = (0..pool.threads()).map(|_| Vec::new()).collect::<Vec<_>>();
[INFO] [stdout] ...   |
[INFO] [stdout] 719 | |                     }
[INFO] [stdout] 720 | |                 }))
[INFO] [stdout]     | |__________________^
[INFO] [stdout]     |
[INFO] [stdout]     = note: trait items always share the visibility of their trait
[INFO] [stdout]     = note: this error originates in the macro `FnOnce` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by lukas-code** [comment]
minimized: [(playground)](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=931f85148423afea62cc9d7eec2fc42c)
```rust
trait Foo {
    fn foo() {
        pub struct Bar;
        impl Bar {
            pub fn bar() {}
        }
    }
}
```

bisects to https://github.com/rust-lang/rust/pull/119505 cc @fmease

@rustbot label -E-needs-bisection -E-needs-mcve S-has-mcve A-visibility

**Comment by fmease** [comment]
I already have PR fixing this, namely #120698. I'd like to improve the approach though, that's why it's waiting on author atm. I will prioritize this PR then.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://github.com/rust-lang/rust/issues/121610)).

Thanks @fmease and @compiler-errors for jumping on this!

@rustbot label -I-prioritize +P-critical

**Comment by fmease** [comment]
Downgrading the priority as discussed [here](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-02-29/near/424042593).

**Comment by fmease** [comment]
Fixed by #122004.
Backported in #122524.
Closing as completed.
